android/src/app/organicmaps/Map.java

```diff
@@ -0,0 +1,381 @@
+  public static final int ANCHOR_LEFT_BOTTOM = (ANCHOR_LEFT | ANCHOR_BOTTOM);
+  public static final int ANCHOR_RIGHT_BOTTOM = (ANCHOR_RIGHT | ANCHOR_BOTTOM);
+
+  // Should correspond to df::TouchEvent::INVALID_MASKED_POINTER from user_event_stream.cpp
+  public static final int INVALID_POINTER_MASK = 0xFF;
+  public static final int INVALID_TOUCH_ID = -1;
+
+  private int mCurrentCompassOffsetX;
+  private int mCurrentCompassOffsetY;
+  private int mBottomWidgetOffsetX;
+  private int mBottomWidgetOffsetY;
+
+  private int mHeight;
+  private int mWidth;
+  private boolean mRequireResize;
+  private boolean mSurfaceCreated;
+  private boolean mSurfaceAttached;
+  private boolean mLaunchByDeepLink;
+  @Nullable
+  private String mUiThemeOnPause;
+  @Nullable
+  private MapRenderingListener mMapRenderingListener;
+  @Nullable
+  private CallbackUnsupported mCallbackUnsupported;
+
+  public Map()
+  {
+    onCreate(false);
+  }
+
+  /**
+   * Moves the map compass using the given offsets.
+   *
+   * @param context     Context.
+   * @param offsetX     Pixel offset from the top. -1 to keep the previous value.
+   * @param offsetY     Pixel offset from the right.  -1 to keep the previous value.
+   * @param forceRedraw True to force the compass to redraw
+   */
+  public void setupCompass(final Context context, int offsetX, int offsetY, boolean forceRedraw)
+  {
+    int x = offsetX < 0 ? mCurrentCompassOffsetX : offsetX;
+    int y = offsetY < 0 ? mCurrentCompassOffsetY : offsetY;
+    int navPadding = UiUtils.dimen(context, R.dimen.nav_frame_padding);
+    int marginX = UiUtils.dimen(context, R.dimen.margin_compass) + navPadding;
+    int marginY = UiUtils.dimen(context, R.dimen.margin_compass_top) + navPadding;
+    nativeSetupWidget(WIDGET_COMPASS,
+        mWidth - x - marginX,
+        y + marginY,
+        ANCHOR_CENTER);
+    if (forceRedraw && mSurfaceCreated)
+      nativeApplyWidgets();
+    mCurrentCompassOffsetX = x;
+    mCurrentCompassOffsetY = y;
+  }
+
+  public static void onCompassUpdated(double north, boolean forceRedraw)
+  {
+    nativeCompassUpdated(north, forceRedraw);
+  }
+
+  /**
+   * Moves the ruler and copyright using the given offsets.
+   *
+   * @param context Context.
+   * @param offsetX Pixel offset from the left.  -1 to keep the previous value.
+   * @param offsetY Pixel offset from the bottom. -1 to keep the previous value.
+   */
+  public void setupBottomWidgetsOffset(final Context context, int offsetX, int offsetY)
+  {
+    int x = offsetX < 0 ? mBottomWidgetOffsetX : offsetX;
+    int y = offsetY < 0 ? mBottomWidgetOffsetY : offsetY;
+    setupRuler(context, x, y);
+    setupAttribution(context, x, y);
+    mBottomWidgetOffsetX = x;
+    mBottomWidgetOffsetY = y;
+  }
+
+  public void onSurfaceCreated(final Context context, final Surface surface, Rect surfaceFrame, int surfaceDpi)
+  {
+    if (nativeIsEngineCreated())
+    {
+      nativeDetachSurface(true);
+    }
```

Comment:
```suggestion
    if (nativeIsEngineCreated())
      nativeDetachSurface(true);
```

We prefer smaller code when possible.